[ONEM-40903] Update LEDControl with the new LED states

srikanth-s-infosys opened the pull request:
[ONEM-40903] Update LEDControl with the new LED states

srikanth-s-infosys committed:
[ONEM-40903] Update LEDControl with the new LED states

srikanth-s-infosys committed:
[ONEM-40903] Update LEDControl.cpp and LEDControl.config

mikolaj-staworzynski-red commented on LEDControl/LEDControl.cpp:
```diff
@@ -264,6 +298,27 @@ namespace WPEFramework
                     state = dsFPD_LED_DEVICE_SOFTWARE_DOWNLOAD_ERROR;
+                } else if (0==strncmp(strLedState.c_str(), FPD_LED_DEVICE_BOOT_IN_PROGRESS, strlen(FPD_LED_DEVICE_BOOT_IN_PROGRESS)) &&
+                                (strlen(strLedState.c_str()) == strlen(FPD_LED_DEVICE_BOOT_IN_PROGRESS)) ){
+                    state = dsFPD_LED_DEVICE_BOOT_IN_PROGRESS;
```
I understand we try to follow the existing implementation, but it looks bad.
Each of the if including the ones we already have there could be replaced by (example):

`if (strLedState == FPD_LED_DEVICE_BOOT_IN_PROGRESS) {
    state = dsFPD_LED_DEVICE_BOOT_IN_PROGRESS;
} else if  ... and similar for all the other cases
}`

Moving forward with that looks like all could be replaced by map:
`int32_t LEDControl::setLEDState(const JsonObject& parameters, JsonObject& response)
{
    LOGINFOMETHOD();
    returnIfParamNotFound(parameters, "state");
    std::string strLedState = parameters["state"].String();
    bool success = false;

    try {
        static const std::unordered_map<std::string, dsFPDLedState_t> stateMap = {
            {FPD_LED_DEVICE_ACTIVE, dsFPD_LED_DEVICE_ACTIVE},
            {FPD_LED_DEVICE_STANDBY, dsFPD_LED_DEVICE_STANDBY},
            {FPD_LED_DEVICE_WPS_CONNECTING, dsFPD_LED_DEVICE_WPS_CONNECTING},
            {FPD_LED_DEVICE_WPS_CONNECTED, dsFPD_LED_DEVICE_WPS_CONNECTED},
            {FPD_LED_DEVICE_WPS_ERROR, dsFPD_LED_DEVICE_WPS_ERROR},
            {FPD_LED_DEVICE_FACTORY_RESET, dsFPD_LED_DEVICE_FACTORY_RESET},
            {FPD_LED_DEVICE_USB_UPGRADE, dsFPD_LED_DEVICE_USB_UPGRADE},
            {FPD_LED_DEVICE_SOFTWARE_DOWNLOAD_ERROR, dsFPD_LED_DEVICE_SOFTWARE_DOWNLOAD_ERROR},
            {FPD_LED_DEVICE_BOOT_IN_PROGRESS, dsFPD_LED_DEVICE_BOOT_IN_PROGRESS},
            {FPD_LED_DEVICE_COLDSTANDBY, dsFPD_LED_DEVICE_COLDSTANDBY},
            {FPD_LED_DEVICE_PSU_FAILURE, dsFPD_LED_DEVICE_PSU_FAILURE},
            {FPD_LED_DEVICE_WPS_SES_OVERLAP, dsFPD_LED_DEVICE_WPS_SES_OVERLAP},
            {FPD_LED_DEVICE_IP_ACQUIRED, dsFPD_LED_DEVICE_IP_ACQUIRED},
            {FPD_LED_DEVICE_NO_IP, dsFPD_LED_DEVICE_NO_IP},
            {FPD_LED_DEVICE_RCU_COMMAND, dsFPD_LED_DEVICE_RCU_COMMAND}
        };

        auto it = stateMap.find(strLedState);
        if (it == stateMap.end()) {
            LOGERR("UNKNOWN state: %s", strLedState.c_str());
            LOGTRACEMETHODFIN();
            return WPEFramework::Core::ERROR_BAD_REQUEST;
        }

        dsFPDLedState_t state = it->second;
        LOGINFO("dsFPSetLEDState state: %s state enum: %d", strLedState.c_str(), state);
        dsError_t err = dsFPSetLEDState(state);
        if (err == dsERR_NONE) {
            success = true;
        } else {
            LOGERR("dsFPSetLEDState returned error %d", err);
            LOGTRACEMETHODFIN();
            return WPEFramework::Core::ERROR_ILLEGAL_STATE;
        }

    } catch (...) {
        LOGERR("Exception in dsFPSetLEDState");
        LOGTRACEMETHODFIN();
        return WPEFramework::Core::ERROR_ILLEGAL_STATE;
    }

    returnResponse(success);
}`

mikolaj-staworzynski-red commented on LEDControl/LEDControl.cpp:
```diff
@@ -202,6 +215,27 @@ namespace WPEFramework
                     case dsFPD_LED_DEVICE_SOFTWARE_DOWNLOAD_ERROR:
                         response["state"] = FPD_LED_DEVICE_SOFTWARE_DOWNLOAD_ERROR;
                         break;
+                    case dsFPD_LED_DEVICE_BOOT_IN_PROGRESS:
```
We could consider a map here:
`uint32_t LEDControl::getLEDState(const JsonObject& parameters, JsonObject& response)
{
    bool success = false;
    try {
        dsFPDLedState_t state;
        dsError_t err = dsFPGetLEDState(&state);
        if (!err) {
            success = true;

            static const std::unordered_map<dsFPDLedState_t, const char*> stateToString = {
                {dsFPD_LED_DEVICE_NONE, FPD_LED_DEVICE_NONE},
                {dsFPD_LED_DEVICE_ACTIVE, FPD_LED_DEVICE_ACTIVE},
                {dsFPD_LED_DEVICE_STANDBY, FPD_LED_DEVICE_STANDBY},
                {dsFPD_LED_DEVICE_WPS_CONNECTING, FPD_LED_DEVICE_WPS_CONNECTING},
                {dsFPD_LED_DEVICE_WPS_CONNECTED, FPD_LED_DEVICE_WPS_CONNECTED},
                {dsFPD_LED_DEVICE_WPS_ERROR, FPD_LED_DEVICE_WPS_ERROR},
                {dsFPD_LED_DEVICE_FACTORY_RESET, FPD_LED_DEVICE_FACTORY_RESET},
                {dsFPD_LED_DEVICE_USB_UPGRADE, FPD_LED_DEVICE_USB_UPGRADE},
                {dsFPD_LED_DEVICE_SOFTWARE_DOWNLOAD_ERROR, FPD_LED_DEVICE_SOFTWARE_DOWNLOAD_ERROR},
                {dsFPD_LED_DEVICE_BOOT_IN_PROGRESS, FPD_LED_DEVICE_BOOT_IN_PROGRESS},
                {dsFPD_LED_DEVICE_COLDSTANDBY, FPD_LED_DEVICE_COLDSTANDBY},
                {dsFPD_LED_DEVICE_PSU_FAILURE, FPD_LED_DEVICE_PSU_FAILURE},
                {dsFPD_LED_DEVICE_WPS_SES_OVERLAP, FPD_LED_DEVICE_WPS_SES_OVERLAP},
                {dsFPD_LED_DEVICE_IP_ACQUIRED, FPD_LED_DEVICE_IP_ACQUIRED},
                {dsFPD_LED_DEVICE_NO_IP, FPD_LED_DEVICE_NO_IP},
                {dsFPD_LED_DEVICE_RCU_COMMAND, FPD_LED_DEVICE_RCU_COMMAND},
            };

            auto it = stateToString.find(state);
            if (it != stateToString.end()) {
                response["state"] = it->second;
            } else {
                LOGERR("Unsupported LEDState %d", state);
                LOGTRACEMETHODFIN();
                return WPEFramework::Core::ERROR_BAD_REQUEST;
            }

        } else {
            LOGERR("dsFPGetLEDState returned error %d", err);
            LOGTRACEMETHODFIN();
            return WPEFramework::Core::ERROR_ILLEGAL_STATE;
        }
    } catch (...) {
        LOGERR("Exception in dsFPGetLEDState");
        LOGTRACEMETHODFIN();
        return WPEFramework::Core::ERROR_ILLEGAL_STATE;
    }

    returnResponse(success);
}
`

srikanth-s-infosys commented:
Hi @mikolaj-staworzynski-red , Addressed your comments, based on your suggested changes modified code and tested with CURL requests and able to get responses successfully. Please review latest changes.

srikanth-s-infosys committed:
[ONEM-40903] Update LEDControl.cpp and LEDControl.config
